### PR TITLE
refactor: wave 3 structural decompositions (db.py, session_tail.py, embed_branch_chunks)

### PR DESCRIPTION
**Pull request by NodeJSmith**
### Wave 3: structural decompositions (db.py, session_tail.py, embed_branch_chunks)

Three structural issues flagged by the codebase health audit (#137, #138, #140), executed against the wave-2 safety-net tests below as the pinning contract.

- **T01 — split `db.py` (#137).** `db.py` transitively imported `numpy`/`fastembed`/`onnxruntime` via `ccrecall.embeddings` just for the embedding-dimension constant, taxing every `get_connection()` caller (including the SessionStart hook hot path) with an avoidable ~200ms+ import. Moved every vec-dependent function (`vec_available`, `chunk_vec_queryable`, `upsert_chunk_vec`, `write_chunk_embedding`, `fetch_branch_messages`, `branch_embedding_coverage`, `_ensure_vec_schema`, and the vec-related constants) into a new `db_vec.py`; `db.py` no longer re-exports them, so all twelve callers were updated to import from `db_vec` directly instead. The one remaining call to `vec_available` inside `_open_connection`'s `load_vec=True` branch is now a deliberate deferred import — the only lazy import in the codebase, justified by CLAUDE.md invariant #3 (hook hot path) taking precedence over the no-lazy-imports rule.
- **T02 — split `session_tail.py` (#140).** The 653-line file mixed path resolution, pending-question detection, tail rendering, and CLI orchestration. Extracted `tail_resolve.py` (path/session resolution) and `tail_pending.py` (pending-question detection + shared formatting helpers, deliberately self-contained with no import back into `session_tail.py` to keep the dependency graph one-directional). `session_tail.py` now only owns tail rendering and CLI entry points. `context_rendering.py`'s imports were updated to pull from the new modules directly.
- **T03 — decompose `embed_branch_chunks` (#138).** This function carries the embedding watermark protocol (CLAUDE.md invariant #3's write-path half) but was 198 lines, too large to review the invariant confidently. Split into named steps (`_prepare_exchange_data`, `_diff_exchanges`, plus embed/write and watermark-check helpers) each under 80 lines, with the clear-first/set-last transaction boundary preserved unchanged. The two backfill `run()` functions were deliberately left unmerged — their structural similarity is superficial (they diverge in tested ways: abort-vs-skip stuck-batch handling, progress/ETA math, SAVEPOINT retry behavior) and unifying them risked silently changing tested behavior.

### Wave 2 safety-net tests

Pinning tests per `rules/common/refactoring-discipline.md`, written before the wave-3 splits above so the refactor was checkable rather than blind:

- `tests/test_transcript_sources.py` — 62 tests covering the transcript-discovery path (project/session dir safety checks, path resolution, symlink handling, glob-injection and path-traversal edge cases, subagent-dir candidate detection) — the surface T02 touches.
- `tests/test_message_ops.py` — 25 tests covering session/message row upsert, `_build_message_row`, `_insert_new_messages`, and `_update_missing_tool_content`, guarding the invariants T01/T03 must preserve.
- `tests/test_cli_smoke.py` — 9 smoke tests across the CLI command surface (`sync-current`, `import`, `status`, `backfill-summaries`, `backfill-tool-content`, `recent`, `search`, `search-messages`, `tail`), so a decomposition that breaks a command's wiring fails a test instead of surfacing at runtime.

### Housekeeping

- Corrected `CLAUDE.md`'s "config.py / db.py split" architecture description to reflect the new `db_vec.py` module (the T01 fixer loop found this stale).
- Corrected a stale docstring claim in `tests/test_context_injection.py` about `db.py` importing fastembed — no longer true post-split.
- Archived `design/specs/012-wave-3-structural-decompositions/` task files (T01–T03) via `cfl archive`; `design.md` retained with `Status: archived`.

## Summary by CodeRabbit

- **New Features**
  - Improved session-tail handling for locating transcripts and identifying unresolved questions.
  - Pending questions are now filtered and formatted more accurately, including prompts from related conversation branches.
  - Added safer transcript discovery across projects and worktrees.

- **Bug Fixes**
  - Improved handling of malformed, missing, unreadable, or unsafe transcript files.
  - Vector and embedding operations are loaded only when needed, improving lightweight database operations.

- **Documentation**
  - Added a design specification documenting the architectural improvements.

### CODE REVIEW

File: tests/test_transcript_sources.py [hunk 1]
```diff
@@ -0,0 +1,701 @@
...
+    def test_self_referential_symlink_does_not_raise(self, tmp_path):
+        # On this Python version, Path.resolve() no longer raises on a
+        # self-referential symlink loop (older versions raised RuntimeError,
+        # which this function is built to suppress) — it just returns the
+        # path unresolved past the loop point. Pin that it doesn't propagate
+        # an exception either way.
+        loop = tmp_path / "loop"
+        loop.symlink_to(loop)
+
+        assert _resolved_path(loop) is not None
```
[the remark this answers is not in the document]
Comment (NodeJSmith) [reply tests/test_transcript_sources.py:185]: Fixed — split into two `@pytest.mark.skipif`-gated tests pinning the version-specific behavior on both sides of the Python 3.13 pathlib change: `test_self_referential_symlink_resolves_without_raising` (3.13+, asserts non-None) and `test_self_referential_symlink_returns_none_pre_313` (<3.13, asserts None). Verified against Python 3.11-3.14 locally.

File: CLAUDE.md [hunk 2]
```diff
@@ -37,7 +37,7 @@ The hard dependency is on undocumented Claude Code internals (the `~/.claude/pro
...
+- **`config.py` / `db.py` / `db_vec.py` split.** `config.py` owns paths, config/settings loading, PID files, and logging setup — zero heavy dependencies (no fastembed/onnxruntime/sqlite_vec). It's the module hooks that don't touch the DB (health checks, the clear-handoff writer, the warm-model process) import instead of pulling the full `db.py` stack onto the hook hot path. `db.py` owns connections and schema application; `db_vec.py` owns every function that touches sqlite-vec, chunk vectors, or embedding constants (`vec_available`, `chunk_vec_queryable`, `upsert_chunk_vec`, `write_chunk_embedding`, `fetch_branch_messages`, `_ensure_vec_schema`, `ensure_vec`, `branch_embedding_coverage`) — kept out of `db.py` so `get_connection()` stays importable without pulling numpy/fastembed/onnxruntime onto the hook hot path. `get_connection()` is a `@contextlib.contextmanager` (`with get_connection(settings) as conn:`) that commits on success, rolls back and closes on exception, and always closes — no connection is left open on an error path.
```
[the remark this answers is not in the document]
Comment (NodeJSmith) [reply CLAUDE.md:40]: Fixed — `ensure_vec` is already in the `db_vec.py` ownership list at CLAUDE.md:40 (added in an earlier commit on this branch, before this review ran). No change needed.

File: design/specs/012-wave-3-structural-decompositions/design.md [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+- `sync_current.py`: `chunk_vec_queryable` → `db_vec`
+- `backfill_tool_content.py`: `VEC_BUSY_TIMEOUT_MS` → `db_vec`
+- `recent_chats.py`: vec-specific imports → `db_vec`
+- `import_conversations.py`: vec-specific imports → `db_vec`
+
+Callers that stay on `ccrecall.db` (non-vec imports): `memory_context.py` (`get_connection`), `memory_setup.py` (`CONTENT_ERROR_VERSION`, `get_connection`), `session_tail.py` (`DEFAULT_PROJECTS_DIR`), `cli/commands.py` (`DEFAULT_PROJECTS_DIR`), `search_query.py` (`escape_like`), `backfill_query.py` (filter constants), `backfill_summaries.py` (`get_connection`).
+
+The `_open_connection` function currently calls `vec_available(conn)` and `_ensure_vec_schema(conn)` inline when `load_vec=True`. After the split, it conditionally imports `db_vec` only inside the `if load_vec:` branch — this is the one acceptable lazy import, guarded by a runtime flag rather than placed at module scope, and it exists to preserve the architectural invariant (no numpy on the hook path). The no-lazy-imports rule's `TYPE_CHECKING` exemption doesn't cover this, but the CLAUDE.md hook-hot-path invariant (#3) takes precedence as the governing constraint.
```
[the remark this answers is not in the document]
Comment (NodeJSmith) [reply design/specs/012-wave-3-structural-decompositions/design.md:86]: Fixed — this is a deliberate, documented exception (design.md's "one acceptable lazy import" note, CLAUDE.md invariant #3), not an accidental policy violation. There's no way to defer the numpy import without some form of deferred import, so restructuring isn't viable here. Instead, updated `tools/check_lazy_imports.py`'s docstring and canonical annotation form to document this as a second approved category alongside circular-import avoidance, so the policy text now matches what the checker already accepts and what `db.py` already does.

File: src/ccrecall/db.py [hunk 4]
```diff
@@ -252,10 +79,28 @@ def _ensure_vec_schema(conn: sqlite3.Connection) -> None:
...
+    the hook hot-path invariant (CLAUDE.md #3) holds.
+    """
+
+    def _prepare_vec(c: sqlite3.Connection) -> bool:
+        # Deferred so this only fires on a version-bump migration, not on
+        # every connection (hook hot-path invariant, CLAUDE.md #3).
+        # lazy-import: db_vec imports numpy transitively
+        from ccrecall.db_vec import vec_available
+
+        return vec_available(c)
```
[the remark this answers is not in the document]
Comment (NodeJSmith) [reply src/ccrecall/db.py:99]: Same underlying issue as the design.md thread — addressed there. Restructuring away from a deferred import isn't achievable without either reintroducing the numpy cost at module-import time or laundering the same lazy import through `importlib` (which would dodge the AST checker without fixing anything). Updated `tools/check_lazy_imports.py`'s docstring to document this as an approved exception category instead, so the policy documentation matches the deliberate design already described in design.md and CLAUDE.md invariant #3.

File: src/ccrecall/tail_resolve.py [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+# sessions — enough to find a timestamp even if the trailing lines are a
+# no-timestamp tool-result burst, without reading a multi-MB file in full.
+_TIMESTAMP_TAIL_LINES = 20
+_BRANCH_HEAD_LINES = 20
+
+
+def transcript_dir(cwd: str, projects_dir: Path = DEFAULT_PROJECTS_DIR) -> Path:
+    """Directory holding this cwd's transcripts (raw slug — see module docstring)."""
+    slug = cwd.replace("\\", "/").replace("/", "-").replace(":", "-").replace(".", "-")
+    return projects_dir / slug
```
[the remark this answers is not in the document]
Comment (NodeJSmith) [reply src/ccrecall/tail_resolve.py:38]: Fixed — replaced the dangling module-docstring reference with an inline description of the slug transform.

File: src/ccrecall/tail_resolve.py [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+            entry = json.loads(line)
+        except json.JSONDecodeError:
+            log.debug("failed to parse transcript tail line: %s", path, exc_info=True)
+            continue
+        ts = entry.get("timestamp")
+        if ts and (latest is None or ts > latest):
+            latest = ts
+    if latest is not None:
+        return latest
+    return Instant.from_timestamp(path.stat().st_mtime).format_iso(unit="millisecond")
```
[the remark this answers is not in the document]
Comment (NodeJSmith) [reply src/ccrecall/tail_resolve.py]: Fixed — guarded both `open()` and `path.stat()` with `try/except OSError`, returning `""` (sorts last under `reverse=True`) on failure instead of letting sort() crash. Added two regression tests.

File: tests/test_cli_smoke.py [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+        ):
+            cmd_tail(None, list_sessions=False, full=False, cwd=None, n=DEFAULT_TAIL_EVENTS)
+        assert exc_info.value.code == 0
+        mock_run.assert_called_once_with(
+            None,
+            list_sessions=False,
+            cwd=None,
+            n=DEFAULT_TAIL_EVENTS,
+            full=False,
+        )
```
[the remark this answers is not in the document]
Comment (NodeJSmith) [reply tests/test_cli_smoke.py:207]: Deferring this one — the test file's own docstring states its deliberate scope ("verify each cyclopts command function calls its owning module's run()... not the underlying logic"). Adding parser-level integration tests through the registered Cyclopts app is valid coverage-gap feedback, but it's new test infrastructure and out of scope for this structural-decomposition PR.

File: tests/test_transcript_sources.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (NodeJSmith) [reply tests/test_transcript_sources.py]: Already addressed — the symlink-loop test was split into two version-gated tests in an earlier commit on this branch (`test_self_referential_symlink_resolves_without_raising` for 3.13+, `test_self_referential_symlink_returns_none_pre_313` for <3.13), which is the same fix the sibling thread on this line already confirmed and resolved. CI is green on all four Python versions.
